# ENCODE WGBS WORKFLOW 

End to end implementation of the [ENCODE](https://www.encodeproject.org/) whole genome bisulfite sequencing workflow described [here](https://www.encodeproject.org/wgbs/). 
Builds off [DNANexus](https://wiki.dnanexus.com/Home) WGBS framework and work from the [Myers lab at the HudsonAlpha Institute for Biotechnology](http://research.hudsonalpha.org/Myers/?page_id=660). Developed in the [Zaitlen Lab](http://zaitlenlab.ucsf.edu/), [Univeristy California at San Francisco](https://www.ucsf.edu/). 

## Overview
------ 
Takes raw whole-genome bisulfite sequencing data (fastq files) and returns the number of methylated CpGs found at each site using the Bismark software package (documentation for this can be found [here](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html)). Performs preliminary methylation analysis based on [MethylKit](https://github.com/al2na/methylKit). 

## Software Requirements
------
Needed to successfuly run bash script implementing workflow  

1. Bowtie 2 read aligner:
    * [Bowtie 2](http://bowtie-bio.sourceforge.net/bowtie2/index.shtml) - used to align raw fastq reads to a genome of choice 

2. Bismark:
    * [Bismark](https://www.bioinformatics.babraham.ac.uk/projects/bismark/) - a software tool developed at the [Babraham Institute](https://www.bioinformatics.babraham.ac.uk/index.html) used to map bisulfite converted reads and determine cytosine methylation states

3. Bisulfite converted genome: 
    * [Raw fasta files](http://hgdownload.cse.ucsc.edu/downloads.html) Needed by Bowtie to map reads. A few steps must be taken to prepare the genome file for use in Bowtie/Bismark.
        1. First, the genome must be built/indexed for Bowtie. This must be done just once before running this pipeline for the first time. This built version can be used for all subsequent runs. Documentation from the next-gen analysis software [Homer](http://homer.ucsd.edu/homer/basicTutorial/mapping.html) provides an excellent tutorial for how to prepare the genome for bowtie.
            * [Illumina iGenome](https://support.illumina.com/sequencing/sequencing_software/igenome.html) provides several pre indexed/built genomes for popular model organisms.
        2. Next, the genome must be bisulfite converted for use in Bismark. The [Bismark documentation](https://rawgit.com/FelixKrueger/Bismark/master/Docs/Bismark_User_Guide.html) provides guidance on this under the **bismark_genome_preparation** section. This step will probably take several hours.
        
 4. Trim Galore: 
    * [Trim Galore adapter trimmer](https://www.bioinformatics.babraham.ac.uk/projects/trim_galore/) - applies quality filters and adapter trimming to raw fastq files. Also developed at the Babraham Institute. Depends on a bunch of software itself 
        1. [Cutadapt](https://cutadapt.readthedocs.io/en/stable/) - removes unwanted sequences from fastq files 
        2. [Fastqc](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) - provides quality information on raw fastq files. Useful for generating visualizations. 
        
5. Samtools: 
    * [Samtools](http://samtools.sourceforge.net/) - needed for manipulation of bam/sam files generated by Bismark. 
    
